File: Python_Ver/draw_figures/plotTestRasters.py
import numpy as np
import matplotlib.pyplot as plt


def plotOneModel(ax, raster, base, color):
    """
    Plots spike raster for a single model.

    Parameters:
    raster : ndarray
        Spike raster data for trials.
    base : float
        Base y-position for the plot.
    color : tuple
        Color of the plot.
    """
    for rasterIdx in range(raster.shape[0]):
        x = np.where(raster[rasterIdx, :], base + rasterIdx * 0.001, np.nan)
        ax.plot(x, "|", color=color, markersize=0.5)


def plotTestRasters(ax, R, Rp, Rsp, neuronIdx, color):
    """
    Plots raster test data for different models.

    Parameters:
    R : ndarray
        Raster data for the real test set.
    Rp : ndarray
        Raster data for the RL predicted test set.
    Rsp : ndarray
        Raster data for the supervised predicted test set.
    neuronIdx : int
        Index of the neuron to plot.
    color : list of tuples
        Colors for different model plots.
    """
    plotTrialNo = min(R.shape[1], 20)

    plotOneModel(ax, R[neuronIdx, :plotTrialNo, :], 1.03, color[0])
    plotOneModel(ax, Rp[neuronIdx, :plotTrialNo, :], 1.00, color[1])
    plotOneModel(ax, Rsp[neuronIdx, :plotTrialNo, :], 0.97, color[2])

    ax.set_ylim([0.97, 1.05])
    plt.box()
    ax.tick_params(axis='y', colors='white', length=0)
    ax.tick_params(axis='x', colors='white', length=0)

File: draw_figures/plotTestRasters.m
function plotTestRasters(R, Rp, Rsp, neuronIdx, color)
plotTrialNo = min(size(R,2), 20);
plotOneModel(squeeze(R(neuronIdx,1:plotTrialNo,:)), 1.03, color(1,:))
plotOneModel(squeeze(Rp(neuronIdx,1:plotTrialNo,:)), 1, color(2,:))
plotOneModel(squeeze(Rsp(neuronIdx,1:plotTrialNo,:)), 0.97, color(3,:))

hold off
box off
set(gca, 'YColor', 'white')
set(gca, 'XColor', 'white')
set(gca, 'TickLength', [0 0])
ylim([0.97 1.05])
end

function plotOneModel(raster, base, color)
for rasterIdx=1:size(raster,1)
    x = raster(rasterIdx,:)*(base+rasterIdx*0.001);
    plot(x, "|", "Color",color, "MarkerSize",0.5); hold on
end
end
